### ISSUE TITLE
question about data augmentation

### ISSUE THREAD

**Issue by panda0307** [gcp/leela-zero#766]
Excuse me: 
Why LZ picks only one random symmetry in parsing data? However, in AGZ paper, they use all **eight** symmetries. I wonder if it is inconsistent. Do I understand correctly? 
Thanks a lot!

**Comment by jkiliani** [comment]
Picking one random symmetry for each position is already generating enough training data for Leela Zero. Lately, significant problems were actually caused in the training by positions from the same game applied to training simultaneously. These sort of correlations weaken the value net, and only recently did the developers here find out about this problem. 

But yes, in theory all eight symmetries are used for parsing, just most likely not in the same training run. The symmetries are more important in MCTS, where they apply some randomness without weakening the engine at all.

**Comment by panda0307** [comment]
Does this problem have a huge impact on the training effect? If it is so, what about the training before? useless?

**Comment by jkiliani** [comment]
Of course the training before wasn't useless, or Leela Zero would never have reached Dan level even with the problem. The training curve would probably have been steeper if this bug had been found earlier. We can expect better progress now.

**Comment by Ttl** [comment]
Training works as described in AGZ paper. All symmetries of the training data are generated and added to the training data set. During training positions are chosen randomly from the data set. Picking all symmetries of the same position to the same mini-batch would cause problems since they are highly correlated and in some positions identical.

**Comment by panda0307** [comment]
I see. Thank you!
@Ttl @jkiliani

**Comment by panda0307** [comment]
How to solve the bug? As I read the code, LZ uses mp.Pipe instead of queue? Then bug has been fixed?
@jkiliani

**Comment by jkiliani** [comment]
@panda0307 Before Thursday, the training code did use a shuffling algorithm to mix the training positions into different batches, but it turned out that it didn't do this well enough. There were still too many positions from the same game in each mini-batch, and that reduced the effectiveness of training. Now, much more shuffling is done, unfortunately that is currently slowing down the training algorithm a lot. @gcp and others are still trying to come up with a solution that shuffles the data effectively without or with very little slowdown.
